### PR TITLE
expression: implement vectorized evaluation for `builtinSHA2Sig`

### PR DESCRIPTION
**Pull request by mmyj**
### What problem does this PR solve? 
implement vectorized evaluation for builtinSHA2Sig, for #12106
### What is changed and how it works?
```
go test -v -benchmem -bench=BenchmarkVectorizedBuiltinEncryptionFunc -run=BenchmarkVectorizedBuiltinEncryptionFunc -args 'builtinSHA2Sig'
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc-4                    1664            698544 ns/op          153104 B/op       4100 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc-4                 1568            798050 ns/op          283517 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#01-4                 1742            678770 ns/op          152785 B/op       4100 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#01-4              1490            772987 ns/op          283192 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#02-4                 1737            693847 ns/op          152977 B/op       4100 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#02-4              1572            780537 ns/op          283389 B/op       5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#03-4                 1462            806317 ns/op          201992 B/op       4100 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#03-4              1300            925791 ns/op          430751 B/op       5121 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#04-4                 1354            866654 ns/op          251259 B/op       4100 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#04-4              1220            984544 ns/op          480013 B/op       5121 allocs/op
PASS
ok      github.com/pingcap/tidb/expression      12.797s
```
### Check List 
Tests 
- Unit test

### CODE REVIEW

File: expression/builtin_encryption_vec.go [hunk 1]
```diff
@@ -108,11 +113,66 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	buf1, err := b.bufAllocator.get(types.ETString, n)
```
Comment (Reminiscent) [inline expression/builtin_encryption_vec.go]: Should be ```buf1, err := b.bufAllocator.get(types.ETInt, n)```

Comment (mmyj) [reply expression/builtin_encryption_vec.go]: I will fix it.

File: expression/builtin_encryption_vec.go [hunk 2]
```diff
@@ -139,11 +145,67 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
 }
 
 func (b *builtinSHA2Sig) vectorized() bool {
-	return false
+	return true
 }
 
 func (b *builtinSHA2Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
```
Comment (SunRunAway) [inline expression/builtin_encryption_vec.go]: Please add a comment like `builtinSHA2Sig.evalString`

File: expression/builtin_encryption_vec.go [hunk 3]
```diff
@@ -139,11 +145,67 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+	result.ReserveString(n)
+	i64s := buf1.Int64s()
+	var (
+		hasher    *hash.Hash
```
Comment (SunRunAway) [inline expression/builtin_encryption_vec.go]: Just use `hasher hash.Hash`, and it works.

File: expression/builtin_encryption_vec.go [hunk 4]
```diff
@@ -139,11 +145,67 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		hasher256 = sha256.New()
+		hasher224 = sha256.New224()
+		hasher384 = sha512.New384()
+		hasher512 = sha512.New()
```
Comment (SunRunAway) [inline expression/builtin_encryption_vec.go]: Could we lazy create these hashers?

Comment (mmyj) [reply expression/builtin_encryption_vec.go]: I will fix it.

File: expression/builtin_encryption_vec.go [hunk 5]
```diff
@@ -139,11 +144,79 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalInt(b.ctx, input, buf1); err != nil {
+		return err
+	}
+	result.ReserveString(n)
+	i64s := buf1.Int64s()
+	buf.MergeNulls(buf1)
```
Comment (SunRunAway) [inline expression/builtin_encryption_vec.go]:
```
// The caller should ensure that all these columns have the same
// length, and data stored in the result column is fixed-length type.
```

`MergeNulls` can only be called if `buf` is fixed-length.

File: expression/builtin_encryption_vec.go [hunk 6]
```diff
@@ -139,11 +144,79 @@ func (b *builtinMD5Sig) vecEvalString(input *chunk.Chunk, result *chunk.Column)
...
+		hasher384 hash.Hash
+		hasher512 hash.Hash
+	)
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		hashLength := i64s[i]
+		var hasher *hash.Hash
```
Comment (SunRunAway) [inline expression/builtin_encryption_vec.go]:
```suggestion
		var hasher hash.Hash
```

We don't have to use a pointer of an interface here. The original value of an interface we state here is null.
See the example https://play.golang.org/p/r1GECExUiRj .

**Comment by SunRunAway** [comment]
@mmyj Please take another look and resolve the comments, Thanks.

File: expression/builtin_encryption_vec.go (hunk 2 above)
Comment (mmyj) [reply expression/builtin_encryption_vec.go]: Should I add a comment like

File: expression/builtin_encryption_vec.go (hunk 3 above)
Comment (mmyj) [reply expression/builtin_encryption_vec.go]: ok, I will fix it.

File: expression/builtin_encryption_vec.go (hunk 5 above)
Comment (mmyj) [reply expression/builtin_encryption_vec.go]: I will fix it.

File: expression/builtin_encryption_vec.go (hunk 6 above)
Comment (mmyj) [reply expression/builtin_encryption_vec.go]: I will fix it.

**Comment by mmyj** [comment]
I remove `MergeNulls` and `*hash.Hash`.

**Comment by mmyj** [comment]
new benchmark result
```
go test -v -benchmem -bench=BenchmarkVectorizedBuiltinEncryptionFunc -run=BenchmarkVectorizedBuiltinEncryptionFunc -args 'builtinSHA2Sig'
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc-4         	    1809	    639286 ns/op	  131240 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc-4      	    1561	    773578 ns/op	  283173 B/op	    5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#01-4      	    1868	    624360 ns/op	  131240 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#01-4   	    1574	    764249 ns/op	  283750 B/op	    5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#02-4      	    1868	    635765 ns/op	  131240 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#02-4   	    1578	    753357 ns/op	  283223 B/op	    5120 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#03-4      	    1555	    761328 ns/op	  180509 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#03-4   	    1328	    926879 ns/op	  431054 B/op	    5121 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-VecBuiltinFunc#04-4      	    1412	    821565 ns/op	  229681 B/op	    3073 allocs/op
BenchmarkVectorizedBuiltinEncryptionFunc/builtinSHA2Sig-NonVecBuiltinFunc#04-4   	    1146	   1014659 ns/op	  480200 B/op	    5121 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	12.684s
```

File: expression/builtin_encryption_vec.go (hunk 2 above)
Comment (Reminiscent) [reply expression/builtin_encryption_vec.go]: > Should I add a comment upon L151? like this
> 
> ```
> // vecEvalString evals SHA2(str, hash_length).
> // See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_sha2
> ```

Yes, it's better to add a comment.

Comment (mmyj) [reply expression/builtin_encryption_vec.go]: Should I add a comment upon other vectorized `vecEvalxxxx` functions?

**Pull request merged**